docs/msi_v2/msi_with_credential_design.md

```diff
@@ -0,0 +1,295 @@
+$certSha256 = [BitConverter]::ToString($sha256.ComputeHash($publicKeyBytes)) -replace "-", ""
+
+Write-Output "🔐 Using SHA-256 Certificate Identifier (kid): $certSha256"
+
+# Step 6: Convert certificate to Base64 for JWT (x5c field)
+$x5c = [System.Convert]::ToBase64String($cert.RawData)
+Write-Output "📜 x5c: $x5c"
+
+# Step 7: Construct the JSON body properly
+$bodyObject = @{
+    cnf = @{
+        jwk = @{
+            kty = "RSA"
+            use = "sig"
+            alg = "RS256"
+            kid = $certSha256  # Use SHA-256 instead of Thumbprint
+            x5c = @($x5c)  # Ensures correct array formatting
+        }
+    }
+    latch_key = $false  # Some VMs need this. Remove in production if unnecessary.
+}
+
+# Convert JSON object to a string
+$body = $bodyObject | ConvertTo-Json -Depth 10 -Compress
+Write-Output "🔹 JSON Payload: $body"
+
+# Step 8: Request MSI credential
+$headers = @{
+    "Metadata" = "true"
+    "X-ms-Client-Request-id" = [guid]::NewGuid().ToString()
+}
+
+$imdsResponse = Invoke-WebRequest -Uri "http://169.254.169.254/metadata/identity/credential?cred-api-version=1.0" `
+    -Method POST `
+    -Headers $headers `
+    -Body $body
+
+$jsonContent = $imdsResponse.Content | ConvertFrom-Json
+
+$regionalEndpoint = $jsonContent.regional_token_url + "/" + $jsonContent.tenant_id + "/oauth2/v2.0/token"
+Write-Output "✅ Using Regional Endpoint: $regionalEndpoint"
+
+# Step 9: Authenticate with Azure
+$tokenHeaders = @{
+    "Content-Type" = "application/x-www-form-urlencoded"
+    "Accept" = "application/json"
+}
+
+$tokenRequestBody = "grant_type=client_credentials&scope=https://management.azure.com/.default&client_id=$($jsonContent.client_id)&client_assertion=$($jsonContent.credential)&client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-bearer"
+
+try {
+    $tokenResponse = Invoke-WebRequest -Uri $regionalEndpoint `
+        -Method POST `
+        -Headers $tokenHeaders `
+        -Body $tokenRequestBody `
+        -Certificate $cert  # Use the full certificate object
+
+    $tokenJson = $tokenResponse.Content | ConvertFrom-Json
+    Write-Output "🔑 Access Token: $($tokenJson.access_token)"
+} catch {
+    Write-Error "❌ Failed to retrieve access token. Error: $_"
+}
+```
+
+## Summary of New APIs on Managed Identity Builder
+
+| API Name                         | Purpose                                                                            |
+|----------------------------------|------------------------------------------------------------------------------------|
+| `GetBindingCertificate()`        | Helper method to get the binding certificate when a credential endpoint exist.     |
+| `GetManagedIdentitySourceAsync()`| Helper method to get the managed identity source.                                  |
+| `WithCorrelationID(GUID id)`     | Sets the correlation id for the managed identity requests (v2 source only)         |    
+
+## Client-Side Telemetry
```

Comment:
This is the first time I'm reading about telemetry. I guess it will be sent to us and we'll have a way to view it?